### PR TITLE
Add UDAP registration metadata validation

### PR DESCRIPTION
**Pull request by vanessuniq**
Summary:
- Adds a strict UDAP Security STU2 registration metadata value object for registration and cancellation payloads.
- Requires explicit allow_insecure_localhost opt-in before SMART/shared HTTP clients accept loopback HTTP URIs or redirects.
- Documents the UDAP DCR metadata boundary, local-development exception, and related architecture decisions.

Validation:
- bundle exec rspec
- bundle exec rubocop --cache false
- bin/docs
- bundle exec jekyll build

### CODE REVIEW

File: lib/safire/protocols/udap_registration_metadata.rb [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+
+      def validate_scope!(metadata)
+        validate_non_blank_string!(metadata, 'scope')
+        return if SCOPE_PATTERN.match?(metadata['scope'])
+
+        fail_validation!(:scope, 'must be a space-delimited OAuth scope string')
+      end
+
+      def validate_registration_fields!(metadata)
+        grants = metadata['grant_types']
```
[the remark this answers is not in the document]
Comment (vanessuniq) [reply lib/safire/protocols/udap_registration_metadata.rb:170]: canonicalizes accepted grant_types before freezing/outputting metadata, so reversed valid input emits the documented stable order.

**Comment by vanessuniq** [comment]
> ## PR #170 — Add UDAP registration metadata validation
> **Overall verdict: Approve.** No must-fix correctness or security bugs found. `UdapRegistrationMetadata` is well-structured: validation ordering is correct, immutability model is sound, the `allow_insecure_localhost` opt-in tightens the existing security posture rather than relaxing it, and test coverage is comprehensive across happy paths, edge cases, cancellation, and extension metadata. Four should-fix items below are advisory.
> 
> **Findings summary (severity order)**
> 
> 1. **should-fix** — `.present?` called without declaring ActiveSupport as a dependency; four new methods will raise `NoMethodError` for non-Rails users.
> 2. **should-fix** — `validate_localhost_policy` is copy-pasted into four separate classes; extract to `URIValidation`.
> 3. **should-fix** — `logo_uri` may be required for all UDAP software statements (including `client_credentials`) per UDAP Security STU2 §4.2.1; the current implementation only requires it for `authorization_code`.
> 4. **should-fix** — `image_uri?` is the only URI helper without a rescue clause.
> 5. **nit** — `absolute_uri?` uses stdlib `URI` while every other URI method in the same class uses `Addressable::URI`.
> 6. **nit** — The spec's inline `reserved_fields` array duplicates the private constant; additions to `RESERVED_FIELDS` in the implementation won't be caught automatically.
> 7. **nit** — The breaking behaviour change (localhost HTTP now requires opt-in) is under "Changed" in the CHANGELOG; a explicit "Breaking" callout or separate heading would help callers upgrade safely.
> 
> Additional findings and fallback items:
> 
> * **should-fix**: `.present?` used without ActiveSupport — breaks non-Rails users (lib/safire/uri_validation.rb:1306)
>   Four new methods call `.present?`, which is defined by ActiveSupport and does not exist in standard Ruby:
> * `strict_https_uri?` (`uri.host.present?`)
> * `localhost_http_uri?` (`uri.host.present?`)
> * `absolute_uri?` (`uri.host.present?`)
> * `validate_non_blank_string!` (`value.present?`)
> 
> If the gem is used in a non-Rails context (e.g. a plain Ruby script or Sinatra), these raise `NoMethodError` at runtime. The tests pass because the test environment loads ActiveSupport transitively, masking the issue.
> 
> Replace with pure-Ruby equivalents:
> 
> * `uri.host.present?` → `uri.host && !uri.host.empty?`
> * `value.present?` → `value.is_a?(String) && !value.strip.empty?`
> 
> Note: the pre-existing `parsed_uri_absolute?` in the same diff already uses `uri.scheme && uri.host` (no `.present?`), so the inconsistency is visible within the same PR.
> 
> * **should-fix**: `validate_localhost_policy` duplicated across four classes [context] (lib/safire/uri_validation.rb)
>   The identical three-line boolean guard appears in `ClientConfig#normalize_localhost_policy`, `HTTPClient#validate_localhost_policy`, `Middleware::HttpsOnlyRedirects#validate_localhost_policy`, and `UdapRegistrationMetadata#validate_localhost_policy`. The `URIValidation` module is already the shared home for URI policy helpers — a single `validate_localhost_policy!(value)` there would eliminate the duplication and ensure a future policy change (e.g. including `'::1'`) is applied consistently.
> * **should-fix**: `logo_uri` may be required for all UDAP software statements, not just `authorization_code` (lib/safire/protocols/udap_registration_metadata.rb:1078)
>   UDAP Security STU2 §4.2.1 lists `logo_uri` in the required software-statement claims table without a grant-type qualifier. The current implementation treats it as an authorization-code-only field, so a `client_credentials` registration that omits `logo_uri` passes validation but may be rejected by a conformant server.
> 
> Verify the normative table at https://hl7.org/fhir/us/udap-security/STU2/registration.html. If `logo_uri` is unconditionally required, move its presence check out of `validate_authorization_fields!` and into `validate_required_fields!` (with HTTPS enforcement remaining conditional on the grant type).
> 
> * **should-fix**: `image_uri?` has no rescue clause (lib/safire/protocols/udap_registration_metadata.rb:1167)
>   `image_uri?` is the only URI helper in this class without a `rescue` block. It is called only after `allowed_registration_uri?` has already parsed the URI successfully, so a second parse failure is extremely unlikely — but it is not impossible with Addressable (e.g. edge cases in path extraction). Adding `rescue Addressable::URI::InvalidURIError; false` makes the method self-contained and consistent with its siblings.
> * **nit**: `absolute_uri?` uses stdlib `URI` while every other URI helper uses `Addressable::URI` (lib/safire/protocols/udap_registration_metadata.rb:1149)
>   All other URI methods in `UdapRegistrationMetadata` (`strict_https_uri?`, `localhost_http_uri?`, `image_uri?`, and the inherited `classify_uri`) parse with `Addressable::URI`. `absolute_uri?` uses `URI.parse` from the standard library, which has different tolerance for international and exotic URIs. Switching to `Addressable::URI.parse` (rescuing `Addressable::URI::InvalidURIError`) makes the class consistent and avoids a subtle dual-parser edge case for contacts that contain unusual scheme or host characters.
> * **nit**: Reserved-fields spec duplicates the private constant — drift-prone (spec/safire/protocols/udap_registration_metadata_spec.rb:1967)
>   The inline `reserved_fields` array in the spec replicates `RESERVED_FIELDS` exactly. Because the constant is `private_constant`, it is not directly reachable from tests, but a future addition to `RESERVED_FIELDS` (e.g. `udap_certifications`) would silently leave the spec without coverage for that field. One option is to expose a read-only class method (`UdapRegistrationMetadata.reserved_fields`) and drive the parameterised test from it.
> * **nit**: Breaking behaviour change should be called out more explicitly in the CHANGELOG [context] (CHANGELOG.md:51)
>   Silently accepting HTTP on localhost was the documented default before this PR (ADR-007); it now requires an explicit opt-in. Any existing caller with `base_url: 'http://localhost:...'` will receive a `ConfigurationError` after upgrading without adding `allow_insecure_localhost: true`. The CHANGELOG entry is under `### Changed` which is technically correct, but the impact is significant. Consider a `### Breaking Changes` subsection or a bold warning line so callers scanning the changelog for upgrade hazards find it immediately.

1. ActiveSupport is a dependency within safire, so it will not raise
2. will investigate
3. `logo_uri` is only required for `authorization_code` per the UDAP spec
4. will investigate
5. will investigate
6. will investigate
7. will investigate
